wom/src/main/scala/wom/types/WomCoproductType.scala

```diff
@@ -0,0 +1,52 @@
+package wom.types
+import wom.WomExpressionException
+import wom.values.{WomCoproductValue, WomOptionalValue, WomValue}
+
+import scala.util.{Failure, Success, Try}
+import mouse.all._
+
+/**
+  * Handles the possibility that a value could be one of the specified types.  At present this is only supported by CWL.
+  */
+case class WomCoproductType(types: List[WomType]) extends WomType {
+
+  /**
+    * Method to be overridden by implementation classes defining a partial function
+    * for the conversion of raw input values to specific implementation class value types.
+    * i.e.  `WomBooleanType` should define a partial function that knows how to
+    * construct `WomBoolean`s for inputs of supported types and contents.  Values for which
+    * the partial function is not defined are assumed to not be convertible to the target type.
+    */
+  override def coercion: PartialFunction[Any, WomValue] = {
+    case wct@WomCoproductValue(tpe, _) if (tpe.equalsType(this).isSuccess) => wct
+
+    //If we can find this type exactly in our coproduct, use that type for the coercion
+    case womValue: WomValue if (types.contains(womValue.womType)) =>
+      val v: Try[WomValue] = womValue.womType.coerceRawValue(womValue)
+      v.get |> (WomCoproductValue(this, _))
+
+    //If we don't have any information, try to coerce this value one by one, stopping at the first successful try
+    case any =>
+      val triedToCoerce: Try[WomValue] = types.map(_.coerceRawValue(any)).reduce(_ orElse _)
```

Comment:
ToL: I have a vague memory that `reduce` throws on an empty list is that correct ? I don't know if it's possible for a `WomCoproductType` to have an empty list of types (definitely wouldn't make sense), but since nothing strictly prevents it just wanted to mention it.

Reply:
:+1:.  it does not make sense as this type only created when number of types > 2 but it is worthwhile to encode this as a type, hence I'm going to use `NonEmptyList` to represent `types`